src/loaders/zarr_utils/utils.ts

```diff
@@ -101,3 +109,130 @@ export function getScale(dataset: OMEDataset | OMEMultiscale, orderTCZYX: TCZYX<
   const scale = scaleTransform.scale.slice();
   return orderByTCZYX(scale, orderTCZYX, 1);
 }
+
+/**
+ * Defines a partial order of zarr arrays based on their size. Specifically:
+ * - If array size x, y, z are all equal, the arrays are equal
+ * - otherwise, if all xyz of `a` are less than or equal to those of `b`, `a` is less than `b` (and vice versa)
+ * - if some xyz is less and some is greater, the arrays are uncomparable
+ */
+function compareZarrArraySize(
+  aArr: NumericZarrArray,
+  aTCZYX: TCZYX<number>,
+  bArr: NumericZarrArray,
+  bTCZYX: TCZYX<number>
+): number | undefined {
+  const diffZ = aArr.shape[aTCZYX[2]] - bArr.shape[bTCZYX[2]];
+  const diffY = aArr.shape[aTCZYX[3]] - bArr.shape[bTCZYX[3]];
+  const aX = aTCZYX[4] > -1 ? aArr.shape[aTCZYX[4]] : 1;
+  const bX = bTCZYX[4] > -1 ? bArr.shape[bTCZYX[4]] : 1;
+  const diffX = aX - bX;
+
+  if (diffZ === 0 && diffY === 0 && diffX === 0) {
+    return 0;
+  } else if (diffZ <= 0 && diffY <= 0 && diffX <= 0) {
+    return -1;
+  } else if (diffZ >= 0 && diffY >= 0 && diffX >= 0) {
+    return 1;
+  } else {
+    return undefined;
+  }
+}
+
+function scaleTransformsAreEqual(aSrc: ZarrSource, aLevel: number, bSrc: ZarrSource, bLevel: number): boolean {
+  const aScale = getScale(aSrc.multiscaleMetadata.datasets[aLevel], aSrc.axesTCZYX);
+  const bScale = getScale(bSrc.multiscaleMetadata.datasets[bLevel], bSrc.axesTCZYX);
+  return aScale[2] === bScale[2] && aScale[3] === bScale[3] && aScale[4] === bScale[4];
+}
+
+/**
+ * Ensures that all scale levels in `sources` are matched up by size. More precisely: enforces that, for any scale
+ * level `i`, the size of zarr array `s[i]` is equal for every source `s`. We accomplish this by removing any arrays
+ * (and their associated OME dataset metadata) which don't match up in all sources.
+ *
+ * Assumes all sources have scale levels ordered by size from largest to smallest. (This should always be true for
+ * compliant OME-Zarr data.)
```

Comment:
Just wondering, Is that confirmed by the v0.4 ome-ngff spec?

Reply:
also, should the comment for this function say that the first source is the target for which all others are trying to match?

Reply:
probably also should say that this function modifies the passed in `sources` array, or even consider having it return the modified array.

Reply:
I did check the spec on this: https://ngff.openmicroscopy.org/latest/#multiscale-md

> Each dictionary in "datasets" MUST contain the field "path", whose value contains the path to the array for this resolution relative to the current zarr group. The "path"s MUST be ordered from largest (i.e. highest resolution) to smallest.

Reply:
The function doesn't really give one source any special treatment, though (per the comment in that file) I only ever use the first `multiscaleMetadata` after this function does its work.

Reply:
I'm all about immutable data and would have implemented this function to return an entirely new `sources` array if that were feasible, but `ZarrSource` is a bit too complex an object to want to clone. It's definitely worth adding a comment though to emphasize "hey, this modifies its input!" (It could also return only the fields it modifies, i.e. new `NumericZarrArray[]` and new `OMEDataset[]`?)

Reply:
Cool, for me it is sufficient to just say it clearly in the comment that the input arg comes out modified